app/service/auth_svc.py

```diff
@@ -88,21 +96,40 @@ async def login_user(self, request):
         :param request:
         :return: the response/location of where the user is trying to navigate
         """
-        data = await request.post()
-        username = data.get('username')
-        password = data.get('password')
-        if self.ldap_config:
-            verified = await self._ldap_login(username, password)
-        else:
-            verified = await self._check_credentials(request.app.user_map, username, password)
-
-        if verified:
-            self.log.debug('%s logging in:' % username)
-            response = web.HTTPFound('/')
-            await remember(request, response, username)
-            raise response
-        self.log.debug('%s failed login attempt: ' % username)
-        raise web.HTTPFound('/login')
+        try:
+            self.log.debug('Using login handler "%s" for login' % self._login_handler.name)
+            await self._login_handler.handle_login(request)
+        except web.HTTPRedirection as http_redirect:
+            raise http_redirect
+        except web.HTTPUnauthorized as http_unauthorized:
+            raise http_unauthorized
+        except Exception as e:
+            self.log.error('Exception when handling login request: %s' % e)
+
+        # Fallback if not already using default login handler
+        if not isinstance(self._login_handler, self.DefaultLoginHandler):
```

Comment:
I'm a bit confused about this flow. If a default handler is configured it seems like we log and suppress the error. Otherwise, if a non-default is configured we fall back to the default and use it. Is that right? What if we didn't fall back?

Reply:
Hmm, it looks like I overlooked the case where an error occurs with the default handler. I'm thinking this will be the flow:

- Use the configured login handler. if it succeeds, then we're done. If it throws a redirect/unauthorized exception, we raise it to the caller and we're done.
- If the configured login handler throws an error that we don't expect (i.e. not redirect or unauthorized):
    - If the configured login handler was the default, then we raise the exception since there's no more backup options
    - If the configured login handler was not the default, then we fall back to the default and raise any exceptions that come up